rama-http/src/service/fs/serve_dir/open_file.rs

```diff
@@ -41,26 +65,95 @@ pub(super) struct FileOpened {
 pub(super) enum FileRequestExtent {
     Full(File, Metadata),
     Head(Metadata),
+    Embedded(Vec<u8>, u64), // Content, original file size
+    EmbeddedHead(u64),      // Content length for HEAD requests on embedded files
+}
+
+impl FileRequestExtent {
+    pub(super) fn reader(self) -> Option<impl AsyncRead + Send + Sync + Unpin> {
+        match self {
+            Self::Head(_) | Self::EmbeddedHead(_) => None,
+            Self::Full(file, _) => Some(Either::A(file)),
+            Self::Embedded(content, _) => Some(Either::B(Cursor::new(content))),
+        }
+    }
+
+    pub(super) fn get_size(&self) -> u64 {
+        match self {
+            Self::Head(meta) | Self::Full(_, meta) => meta.len(),
+            Self::Embedded(_, original_size) => *original_size,
+            Self::EmbeddedHead(size) => *size,
+        }
+    }
+}
+
+pub(super) fn open_file_embedded(
+    base: &Dir,
+    mut path_to_file: PathBuf,
+    req: &Request,
+    negotiated_encodings: &[QualityValue<Encoding>],
+    range_header: Option<&str>,
+    buf_chunk_size: usize,
+) -> io::Result<OpenFileOutput> {
+    // Check if this is a directory request (empty path or directory path)
+    // If so, try to serve index.html
+    if path_to_file.as_os_str().is_empty() || path_to_file.to_str() == Some(".") {
+        path_to_file = PathBuf::from("index.html");
```

Comment:
shouldn't this behavior depend on the serve mode? Only when append index mode is used or w/e. As right now I think you are always doing it? Check how the regular FS based approach does it.